### PR TITLE
[TC classifier][XDP] Make sure the manager plays well with other services that might be using the clsact qdisc

### PR DESCRIPTION
**Pull request by Gui774ume**
### What does this PR do?

This PR ensures that the manager plays well with other services that might be using the `clsact` qdisc. The point here is to make sure that it won't touch any BPF filters that don't need to be touched, as well as perform a proper cleanup on start up if a previous manager was abruptly killed.

This PR cleans up the netlink socket cache to prevent leaking network namespaces. Important comments were added when actions are expected from the user of the manager.

```golang
type Probe struct {
        [...]

	// IfIndexNetns - (TC Classifier & XDP) Network namespace in which the network interface lives. If this value is
	// provided, then IfIndexNetnsID is required too.
	// WARNING: it is up to the caller of "Probe.Start()" to close this netns handle. Failing to close this handle may
	// lead to leaking the network namespace. This handle can be safely closed once "Probe.Start()" returns.
	IfIndexNetns uint64

	// IfIndexNetnsID - (TC Classifier & XDP) Network namespace ID associated of the IfIndexNetns handle. If this value
	// is provided, then IfIndexNetns is required too.
	// WARNING: it is up to the caller of "Probe.Start()" to call "manager.CleanupNetworkNamespace()" once the provided
	// IfIndexNetnsID is no longer needed. Failing to call this cleanup function may lead to leaking the network
	// namespace. Remember that "manager.CleanupNetworkNamespace()" will close the netlink socket opened with the netns
	// handle provided above. If you want to start the probe again, you'll need to provide a new valid netns handle so
	// that a new netlink socket can be created in that namespace.
	IfIndexNetnsID uint32

        [...]
}
```

```golang
// CleanupNetworkNamespace - Cleans up all references to the provided network namespace within the manager. This means
// that any TC classifier or XDP probe in that network namespace will be stopped and all opened netlink socket in that
// namespace will be closed.
// WARNING: Don't forget to call this method if you've provided a IfindexNetns and IfindexNetnsID to one of the probes
// of this manager. Failing to call this cleanup function may lead to leaking the network namespace. Only call this
// function when you're sure that the manager no longer needs to perform anything in the provided network namespace (or
// else call NewNetlinkConnection first).
func (m *Manager) CleanupNetworkNamespace(netnsID uint32) error {
        [...]
}
```

Two new options were also added to the TC classifier probe type:

```golang
type Probe struct {
        [...]

	// TCFilterPrio - (TC classifier) defines the priority of the classifier added to the clsact qdisc. Defaults to 50.
	TCFilterPrio uint16

	// TCFilterProtocol - (TC classifier) defines the protocol to match in order to trigger the classifier. Defaults to
	// ETH_P_ALL.
	TCFilterProtocol uint16

        [...]
}
```

### Describe how to test your changes

The TC example was updated to test the new changes.

### CODE REVIEW

**Review by safchain — APPROVED** [review]

File: manager.go [hunk 1]
```diff
@@ -1569,23 +1585,87 @@ func (m *Manager) sanityCheck() error {
...
+
+// CleanupNetworkNamespace - Cleans up all references to the provided network namespace within the manager. This means
+// that any TC classifier or XDP probe in that network namespace will be stopped and all opened netlink socket in that
+// namespace will be closed.
+// WARNING: Don't forget to call this method if you've provided a IfIndexNetns and IfIndexNetnsID to one of the probes
+// of this manager. Failing to call this cleanup function may lead to leaking the network namespace. Only call this
+// function when you're sure that the manager no longer needs to perform anything in the provided network namespace (or
+// else call NewNetlinkSocket first).
+func (m *Manager) CleanupNetworkNamespace(nsID uint32) error {
+	var err error
```
Comment (brycekahle) [inline manager.go:1640]: Does this need to grab the `stateLock` mutex?

File: manager.go [hunk 2]
```diff
@@ -10,9 +10,12 @@ import (
 	"sync"
 	"time"
 
+	"github.com/vishvananda/netns"
```
Comment (brycekahle) [inline manager.go:13]: nit: cleanup import formatting

File: manager.go [hunk 3]
```diff
@@ -223,28 +226,44 @@ type Options struct {
...
+	TCFilterCount map[int]int
 }
 
-// netlinkCacheValue - (TC classifier programs only) Netlink socket and qdisc object used to update the classifiers of
-// an interface
-type netlinkCacheValue struct {
-	rtNetlink     *tc.Tc
-	schedClsCount int
+// NewNetlinkSocket - Returns a new NetlinkSocket instance
+func NewNetlinkSocket(nsHandle uint64, nsID uint32) (*NetlinkSocket, error) {
```
Comment (brycekahle) [inline manager.go:237]: `nsID` is unused in this function

File: manager.go [hunk 4]
```diff
@@ -1569,23 +1585,87 @@ func (m *Manager) sanityCheck() error {
...
-	var cacheEntry netlinkCacheValue
-	var err error
-	// Open a netlink socket for the requested namespace
-	cacheEntry.rtNetlink, err = tc.Open(&tc.Config{
-		NetNS: int(netns),
-	})
+// NewCachedNetlinkSocket - TC classifiers are attached by creating a qdisc on the requested interface. A netlink socket
+// is required to create a qdisc (or to attach an XDP program to an interface). Since this socket can be re-used for
+// multiple probes, instantiate the connection at the manager level and cache the netlink socket.
+func (m *Manager) NewCachedNetlinkSocket(nsHandle uint64, nsID uint32) (*NetlinkSocket, error) {
```
Comment (brycekahle) [inline manager.go:1591]: Could you clarify what type of ID the `nsID` value is? Is it the netlink-provided ID, the inode number of the network namespace, something else?

File: probe.go [hunk 5]
```diff
@@ -413,14 +447,19 @@ func (p *Probe) init() error {
...
-		}
-		p.Ifindex = int32(inter.Index)
+	// resolve netns ID from netns handle
+	if p.IfIndexNetns == 0 && p.IfIndexNetnsID != 0 || p.IfIndexNetns != 0 && p.IfIndexNetnsID == 0 {
+		return fmt.Errorf("both IfIndexNetns and IfIndexNetnsID are required if one is provided (IfIndexNetns: %d IfIndexNetnsID: %d)", p.IfIndexNetns, p.IfIndexNetnsID)
+	}
+
+	// set default TC classifier priority
+	if p.TCFilterPrio == 0 {
+		p.TCFilterPrio = 50
```
[the remark this answers is not in the document]
Comment (Gui774ume) [reply probe.go:457]: On it 🐌

File: manager.go (hunk 1 above)
Comment (Gui774ume) [reply manager.go:1640]: Good catch, and yes definitely. ✅

File: manager.go (hunk 2 above)
Comment (Gui774ume) [reply manager.go:13]: 👍🏻

File: manager.go (hunk 3 above)
Comment (Gui774ume) [reply manager.go:237]: 👍🏻

File: manager.go (hunk 4 above)
Comment (Gui774ume) [reply manager.go:1591]: Hey 👋🏻 
This is the ID retrieved in `net.ns.inum` (`net` => `struct net`, `ns` => `struct ns_common`, `inum` => `unsigned int`). I'm not exactly sure what you mean by `inode number of the network namespace` because afaik multiple files (thus inodes) could hold a reference to a network namespace ... but maybe there is another inode number I don't know about.
TL;DR I guess this is the netlink-provided ID, if that ID matches the one held by the structures described above 😄 
I added something in the function comment :)

Comment (brycekahle) [reply manager.go:1591]: That actually is NOT the netlink ID. We call it the inode number in code (https://github.com/DataDog/datadog-agent/blob/e6a153e8812334792fc6489a20fbdb2cae30f05d/pkg/process/util/netns.go#L129), but you can view it from the filesystem with `sudo stat /proc/1/ns/net`:
`File: /proc/1/ns/net -> net:[4026531992]`

Comment (brycekahle) [reply manager.go:1591]: Docs are limited, but I found this from `ip-netns(8)`
```
ip netns set NAME NETNSID - assign an id to a peer network
       namespace

              This command assigns a id to a peer network namespace.
              This id is valid only in the current network namespace.
              If the keyword "auto" is specified an available nsid will
              be chosen.  This id will be used by the kernel in some
              netlink messages. If no id is assigned when the kernel
              needs it, it will be automatically assigned by the kernel.
              Once it is assigned, it's not possible to change it.
```
The netlink ID only seems to be used when obtaining data over netlink. It is sent as ancillary data. If you want to lookup the mapping between the netlink ID and the inode/other ID, you have to issue another netlink request.

**Review by lebauce — APPROVED** [review]
